packages/regenesis-surgery/scripts/surgery.ts

```diff
@@ -0,0 +1,346 @@
+    REGEN__L2_PROVIDER_URL
+  )
+
+  // Create an empty object that represents the new genesis state
+  // We're going to move items from the dump into this genesis state
+  const genesis: ChainState = {}
+
+  // Step 1. Transfer the state of each precompiled contract.
+  for (const [address, account] of Object.entries(dump.accounts)) {
+    if (address.startsWith('0x00000000000000000000000000000000000000')) {
+      genesis[address] = account
+      delete dump.accounts[address]
+    }
+  }
+
+  // Step 2. Transfer over each EOA address and turn it into a normal EOA.
+  for (const [address, account] of Object.entries(dump.accounts)) {
+    if (EOA_CODE_HASHES.includes(account.codeHash)) {
+      genesis[address] = {
+        balance: account.balance,
+        nonce: account.nonce,
+        root: DEFAULT_STORAGE_ROOT,
+        codeHash: DEFAULT_CODE_HASH,
+      }
+      delete dump.accounts[address]
+    }
+  }
+
+  /* --- BEGIN UNISWAP SURGERY SECTION --- */
+
+  // Set up the uniswap factory contract reference
+  const UniswapV3Factory = new ethers.Contract(
+    UNISWAP_V3_FACTORY_ADDRESS,
+    UNISWAP_V3_FACTORY_ABI,
+    l2Provider
+  )
+
+  // Step 3. (UNISWAP) Fix the UniswapV3Factory `owner` address.
+  const oldOwnerSlot = toHex32(0)
+  const newOwnerSlot = toHex32(3)
+  dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[newOwnerSlot] =
+    dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldOwnerSlot]
+  delete dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldOwnerSlot]
+
+  // Step 4. (UNISWAP) Fix the UniswapV3Factory `feeAmountTickSpacing` mapping.
+  const feeEvents = await UniswapV3Factory.queryFilter(
+    UniswapV3Factory.filters.FeeAmountEnabled()
+  )
+  for (const event of feeEvents) {
+    const oldSlotKey = getMappingKey([event.args.fee], 1)
+    const newSlotKey = getMappingKey([event.args.fee], 4)
+    dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[newSlotKey] =
+      dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldSlotKey]
+    delete dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldSlotKey]
+  }
+
+  // Step 5. (UNISWAP) Figure out the old and new pool addresses.
+  const pools: {
+    [oldAddress: string]: {
+      newAddress: string
+      token0: string
+      token1: string
+      fee: ethers.BigNumber
+    }
+  } = {}
+  const poolEvents = await UniswapV3Factory.queryFilter(
+    UniswapV3Factory.filters.PoolCreated()
+  )
+  for (const event of poolEvents) {
+    const oldPoolAddress = getUniPoolAddress(
+      event.args.token0,
+      event.args.token1,
+      event.args.fee,
+      UNISWAP_V3_FACTORY_ADDRESS,
+      OLD_UNISWAP_POOL_INITCODE_HASH
+    )
+    const newPoolAddress = getUniPoolAddress(
+      event.args.token0,
+      event.args.token1,
+      event.args.fee,
+      UNISWAP_V3_FACTORY_ADDRESS,
+      NEW_UNISWAP_POOL_INITCODE_HASH
+    )
+    // TODO: Throw an error if any of the pool addresses don't exist in the old state.
+    if (oldPoolAddress in dump.accounts) {
```

Comment:
Are the addresses in `dump.accounts` guaranteed to be all lowercase addresses? Just confirming since `getUniPoolAddress` returns lowercase, not checksummed